src/fcntl.rs

```diff
@@ -193,42 +196,52 @@ feature! {
 
 // The conversion is not identical on all operating systems.
 #[allow(clippy::useless_conversion)]
-pub fn open<P: ?Sized + NixPath>(path: &P, oflag: OFlag, mode: Mode) -> Result<RawFd> {
-    let fd = path.with_nix_path(|cstr| {
-        unsafe { libc::open(cstr.as_ptr(), oflag.bits(), mode.bits() as c_uint) }
+pub fn open<P: ?Sized + NixPath>(
+    path: &P,
+    oflag: OFlag,
+    mode: Mode,
+) -> Result<OwnedFd> {
+    let fd = path.with_nix_path(|cstr| unsafe {
+        libc::open(cstr.as_ptr(), oflag.bits(), mode.bits() as c_uint)
     })?;
 
-    Errno::result(fd)
+    Errno::result(fd).map(|fd| unsafe { OwnedFd::from_raw_fd(fd) })
 }
 
 // The conversion is not identical on all operating systems.
 #[allow(clippy::useless_conversion)]
 #[cfg(not(target_os = "redox"))]
-pub fn openat<P: ?Sized + NixPath>(
-    dirfd: RawFd,
+pub fn openat<Fd: AsFd, P: ?Sized + NixPath>(
+    dirfd: &Fd,
     path: &P,
     oflag: OFlag,
     mode: Mode,
-) -> Result<RawFd> {
-    let fd = path.with_nix_path(|cstr| {
-        unsafe { libc::openat(dirfd, cstr.as_ptr(), oflag.bits(), mode.bits() as c_uint) }
+) -> Result<OwnedFd> {
+    let fd = path.with_nix_path(|cstr| unsafe {
+        libc::openat(
+            dirfd.as_fd().as_raw_fd(),
+            cstr.as_ptr(),
+            oflag.bits(),
+            mode.bits() as c_uint,
+        )
     })?;
-    Errno::result(fd)
+
+    Errno::result(fd).map(|fd| unsafe { OwnedFd::from_raw_fd(fd) })
 }
 
 #[cfg(not(target_os = "redox"))]
-pub fn renameat<P1: ?Sized + NixPath, P2: ?Sized + NixPath>(
-    old_dirfd: Option<RawFd>,
+pub fn renameat<Fd1: AsFd, P1: ?Sized + NixPath, Fd2: AsFd, P2: ?Sized + NixPath>(
+    old_dirfd: &Fd1,
```

Comment:
What was the rationale for using options anyway when you could use `libc::AT_FDCWD`? The intended usage is that people will pass in a `nix::AT_FDCWD` instead of None.

Reply:
It's a lot Rustier.  In particular, most functions that accept `RawFd` can't accept `libc::AT_FDCWD` .  Putting `Option` in the signature makes it clear which do.  If anything, the `Option` is even more helpful now that we have I/O safety.

Reply:
>  In particular, most functions that accept RawFd can't accept libc::AT_FDCWD

true

But one drawback of using `Option<Fd>` is that one has to specify 2 concrete types when passing `None`:

```rs
renameat::<?, ?>(None, oldpath, None, newpath)
```